File: src/main/java/quickus/core/ExecutableStep.java
package quickus.core;

import quickus.types.StepType;

import java.util.Collection;
import java.util.LinkedList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ExecutableStep {
    protected static String paramNameRegex = "\\$\\{%s\\}";
    protected static String param = "\\$\\{(.*?)\\}";
    protected static Pattern paramPattern = Pattern.compile(param);

    protected final QuickusStep step;
    protected final StepType stepType;
    protected final QuickusContext ctx;
    protected final Object[] params;
    protected final Collection<String> paramNames = new LinkedList<>();

    public ExecutableStep(StepType stepType, QuickusStep step, QuickusContext ctx, Object[] params) {
        this.stepType = stepType;
        this.step = step;
        this.ctx = ctx;
        this.params = params;
    }

    public QuickusStep step() {
        return step;
    }

    public void setParamsToContext(String sentence) {
        Matcher matcher = paramPattern.matcher(sentence);
        int i = 0;
        while (matcher.find()) {
            String paramName = matcher.group(1);
            paramNames.add(paramName);
            if ((i+1) <= params.length) {
                ctx.setParameter(paramName, params[i++]);
            } else {
                throw new IllegalArgumentException("You've forgotten to pass the value for parameter [" + paramName + "]");
            }
        }
    }

    public void report(String sentence) {
        String parsedSentence = parseSentence(sentence);
        // TODO pass to reports
        System.out.println(stepType + " " + parsedSentence);
    }

    public String parseSentence(String sentence) {
        for (String paramName : paramNames) {
            String regex = String.format(paramNameRegex, paramName);
            sentence = sentence.replaceAll(regex, ctx.getParameter(paramName, Object.class).toString());
        };
        return sentence;
    }

    public void execute(String sentence) {
        setParamsToContext(sentence);
        report(sentence);
        step.run(ctx);
    }
}

File: src/main/java/quickus/core/ExecutableStep.kt
package quickus.core

import quickus.types.StepType
import java.util.*
import java.util.regex.Pattern

class ExecutableStep(
    private val stepType: StepType,
    private val step: QuickusStep,
    val ctx: QuickusContext,
    private val params: Array<out Any>
) {
    companion object {
        private var paramNameRegex = "\\$\\{%s\\}"
        private var param = "\\$\\{(.*?)\\}"
        private var paramPattern = Pattern.compile(param)
    }

    private val paramNames: MutableCollection<String> = LinkedList()
    fun step(): QuickusStep {
        return step
    }

    private fun setParamsToContext(sentence: String) {
        val matcher = paramPattern.matcher(sentence)
        var i = 0
        while (matcher.find()) {
            val paramName = matcher.group(1)
            paramNames.add(paramName)
            if (i + 1 <= params.size) {
                ctx.setParameter(paramName, params[i++])
            } else {
                throw IllegalArgumentException("You've forgotten to pass the value for parameter [$paramName]")
            }
        }
    }

    private fun report(sentence: String) {
        val parsedSentence = parseSentence(sentence)
        // TODO pass to reports
        println("$stepType $parsedSentence")
    }

    private fun parseSentence(sentence: String): String {
        var evaluatedSentence = sentence
        for (paramName in paramNames) {
            val regex = String.format(paramNameRegex, paramName)
            evaluatedSentence = evaluatedSentence.replace(regex.toRegex(), ctx.getParameter(paramName, Any::class.java).toString())
        }
        return evaluatedSentence
    }

    fun execute(sentence: String) {
        setParamsToContext(sentence)
        report(sentence)
        step.run(ctx)
    }
}